### PR TITLE
Add search command (WIP)

### PR DESCRIPTION
**Pull request by chrisforrette**
This is a work-in-progress PR to add an interactive `search` command to `tink`.

**Usage:**

* `./bin/tink.js search`: Fully interactive package search
* `./bin/tink.js search "some search term"`: Auto-execute a search for the passed in terms.

**Added dependencies:**

* `ink-select-input@2.0.0`
* `ink-spinner@2.0.0`
* `ink-text-input@2.0.0`

**Todos:**
- [x] Iron out input focus. There is currently a focus conflict between the text input for search terms and the select input for selecting matched packages for install in that they are both focused on, and when you hit "Enter" in the latter, both of them receive a submit action.
- [x] Handle CLI arguments. Currently all options from `lib/common-opts.js` are being passed into the `Search` component but aren't addressed at all. We may not need all/any of those options, and we should add [options to be passable to NPM search](https://www.npmjs.com/package/libnpmsearch#opts-for-libnpmsearch-commands).
- [ ] Integrate `PackageView`. This is being built by @larsgw and might happen after this PR, and would display package details when highlighting packages in search results.

### CODE REVIEW

File: bin/tink.js [hunk 1]
```diff
@@ -6,6 +6,10 @@ const CMDS = new Map([
   ['sh', require('../lib/commands/shell.js')],
   ['shell', require('../lib/commands/shell.js')],
   ['org', require('../lib/commands/org.jsx')],
+  ['s', require('../lib/commands/search.jsx')],
+  ['se', require('../lib/commands/search.jsx')],
+  ['find', require('../lib/commands/search.jsx')],
+  ['search', require('../lib/commands/search.jsx')],
```
Comment (zkat) [inline bin/tink.js]: I don't think we need this duplication anymore since I made some changes to the way commands get loaded. Wanna try without?

File: lib/commands/search.jsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  options: Object.assign(require('../common-opts.js'), {}),
+  handler: search
+}
+
+async function search (argv) {
+  let unmount
+
+  const onError = () => {
+    unmount()
+    process.exit(1)
```
Comment (zkat) [inline lib/commands/search.jsx]: is this the expected pattern for `ink`? It's so strange.

File: lib/components/search.jsx [hunk 3]
```diff
@@ -0,0 +1,202 @@
+'use strict'
```
Comment (zkat) [inline lib/components/search.jsx:1]: I really like the idea of putting things in `lib/components/`!

File: lib/components/search.jsx [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+  return null
+}
+
+const PackageSelectIndicator = ({ isSelected }) => {
+  if (!isSelected) {
+    return ' ';
+  }
+
+  return <Color green>{`${figures.pointer} `}</Color>
```
Comment (zkat) [inline lib/components/search.jsx]: you should conditionalize this on `opts.unicode`. We don't always display special characters, and a lot of terminals just plain can't handle them. Or just skip `figures` since it's yet another dependency.

File: lib/commands/search.jsx [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+const SearchCommand = module.exports = {
+  command: 'search',
+  aliases: ['s', 'se', 'find'],
+  describe: 'Search the NPM registry',
+  builder (y) {
+    return y.help().alias('help', 'h')
+      .options(SearchCommand.options)
+  },
+  options: Object.assign(require('../common-opts.js'), {}),
```
Comment (chrisforrette) [inline lib/commands/search.jsx]: @zkat [What you were saying over here](https://npm.community/t/tink-new-search-ui/3087/22?u=chrisforrette) makes sense about passing the whole options object to `libnpm.search`—so just to confirm, I don't need `common-opts.js` at all for the search command, right? So I can change this to a `figgy-pudding` object with [all the `libnpmsearch` opts](https://www.npmjs.com/package/libnpmsearch#opts-for-libnpmsearch-commands)?

File: lib/components/search.jsx [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+        maintenance,
+        popularity,
+        quality
+      }
+    }
+  } = value
+
+  const m = formatPackageScore(maintenance)
+  const p = formatPackageScore(popularity)
+  const q = formatPackageScore(quality)
```
Comment (zkat) [inline lib/components/search.jsx:217]: How do you feel about the `m/p/q` formatting? Do you think it's helpful to people to display them like that?

File: lib/components/search.jsx [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+  const items = matches.map(match => ({
+    value: match,
+    label: match
+  }))
+
+  return <SelectInput
+    indicatorComponent={PackageSelectIndicator}
+    items={items}
+    itemComponent={PackageItem}
+    onSelect={({ value }) => onSelect(value)} />
```
Comment (zkat) [inline lib/components/search.jsx]: JSX here is so delightful

File: package.json [hunk 8]
```diff
@@ -62,7 +62,10 @@
     "graceful-fs": "^4.1.11",
     "ini": "^1.3.5",
     "ink": "^0.5.1",
+    "ink-select-input": "^2.0.0",
+    "ink-spinner": "^2.0.0",
     "ink-table": "^1.0.3",
+    "ink-text-input": "^2.0.0",
     "jsx-transform": "^2.4.0",
```
Comment (zkat) [inline package.json:107]: If you're gonna use `figures`, please make sure it ends up in our dependencies!

File: lib/components/search.jsx [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+  onSelectPackage (pkg) {
+    this.install(pkg)
+  }
+}
+
+const SearchInput = ({ onChange, onSubmit, terms }) => {
+  return <div>
+    <Color grey bold>Find a package: </Color>
+    <TextInput value={terms} onChange={onChange} onSubmit={onSubmit} />
+  </div>
```
Comment (zkat) [inline lib/components/search.jsx:168]: For some reason, if I erase the textinput data and press "Enter" it does BOTH a search and an install (and closes the process). Or was I testing an older version?

**Review by zkat — CHANGES_REQUESTED** [review]

File: lib/commands/search.jsx (hunk 2 above)
Comment (chrisforrette) [reply lib/commands/search.jsx]: @zkat I actually borrowed this pretty much verbatim from the `emma-cli`:

https://github.com/maticzav/emma-cli/blob/master/src/index.js#L51

Definitely a little weird, but I think it makes sense...

File: lib/components/search.jsx (hunk 6 above)
Comment (chrisforrette) [reply lib/components/search.jsx:217]: @zkat I think it's pretty vague—I definitely didn't know what it was until I put it together after seeing the `sortBy` option in `libnpmsearch`. It's nice having a little more information in a one-liner, but we're definitely limited in what we can fit. Maybe we could change the one-liner to package name, latest version, maybe publishing user? And once `<PackageView />` is in the mix, these handy details can be displayed with a little more context.

What do you think?

File: package.json (hunk 8 above)
Comment (chrisforrette) [reply package.json:107]: Woops—it's a dependency of `ink-spinner` and I was tinkering... I'll add it!

File: lib/components/search.jsx (hunk 9 above)
Comment (chrisforrette) [reply lib/components/search.jsx:168]: Yea! This is that weird input focus thing I mentioned in my Todos list. I saw that the `ink-select-input` has an explicit `focus` boolean prop, and I just realized `ink-text-input` has the same prop undocumented. I just need to add some logic to juggle these intuitively...

File: lib/components/search.jsx (hunk 6 above)
Comment (zkat) [reply lib/components/search.jsx:217]: How about:
```
Search: Foo
Results (maintenance, popularity, quality):
  -> foo@1.2.3 @zkat (50%/60%/75%)
```

Comment (zkat) [reply lib/components/search.jsx:217]: screenshot_20181108-062110

npms also has a neat overall score, and we could display the full breakdown on focus, as you suggested.

File: lib/commands/search.jsx [hunk 10]
```diff
@@ -0,0 +1,38 @@
+'use strict'
```
Comment (legodude17) [inline lib/commands/search.jsx]: Why is this file named as `search.jsx`? AFAICT it has no JSX in it.

File: lib/components/search.jsx [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+  Fragment,
+  Text
+} = require('ink')
+const SelectInput = require('ink-select-input')
+const Spinner = require('ink-spinner')
+const TextInput = require('ink-text-input')
+const libnpm = require('libnpm')
+
+
+const MAX_RESULTS = 10
```
Comment (legodude17) [inline lib/components/search.jsx]: Shouldn't this be configurable?

File: lib/components/search.jsx [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+      <PackageSelector matches={matches} onSelect={onSelect} />
+    </div>
+  }
+
+  return null
+}
+
+const PackageSelectIndicator = ({ isSelected }) => {
+  if (!isSelected) {
+    return ' ';
```
Comment (legodude17) [inline lib/components/search.jsx]: Shouldn't this be two spaces so that it is the same length as when it is selected?
```suggestion
    return '  ';
```

File: lib/commands/search.jsx [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+
+const { h, render } = require('ink')
+const { Search } = require('../components/search.jsx')
```
Comment (legodude17) [inline lib/commands/search.jsx]: Why don't you assign `Search` directly to `module.exports` in `../components/search.jsx`?

**Review by legodude17 — COMMENTED** [review]
So excited for this! 🎉

File: lib/commands/search.jsx (hunk 13 above)
Comment (zkat) [reply lib/commands/search.jsx]: this is a bit nitpicky, legodude

File: lib/components/search.jsx (hunk 11 above)
Comment (zkat) [reply lib/components/search.jsx]: I don't really care to make something like this configurable. Just pick good defaults imo

File: lib/commands/search.jsx (hunk 10 above)
Comment (zkat) [reply lib/commands/search.jsx]: Probably didn't get around to actually using it here: https://github.com/npm/tink/pull/11/files/ef5c1c91aeff46b8a4d4e3e76f092f4e58b898fa#diff-08239d674f7fed0d6d8f684e18ce64d0R33

but there might well be some JSX in the future in here.

File: lib/commands/search.jsx (hunk 13 above)
Comment (legodude17) [reply lib/commands/search.jsx]: Honestly I was mostly wondering if there were plans to export more things from `../components/search.jsx`.

File: lib/components/search.jsx (hunk 12 above)
Comment (larsgw) [reply lib/components/search.jsx]: I just tested this and... it doesn't work? If I change it to a single `#` the indentation is one char (no spaces), as expected. However, when I keep it as a single space it shows two spaces and when I add the second it shows three... There are some weird things going on with spacing here and there in either ink or jsx.

Comment (legodude17) [reply lib/components/search.jsx]: Huh. That's really interesting.

Comment (zkat) [reply lib/components/search.jsx]: I guess it tried to simulate html in that aspect lol

File: lib/commands/search.jsx (hunk 13 above)
Comment (zkat) [reply lib/commands/search.jsx]: possibly in the future, but not as of yet :)

Maybe when we have workspace support.

**Comment by zkat** [comment]
FYI you'll need to accept https://github.com/npm/tink/pull/22 before I can merge this.

File: lib/components/search.jsx (hunk 6 above)
Comment (chrisforrette) [reply lib/components/search.jsx:217]: @zkat Yea, I think a mini legend works great.

File: lib/commands/search.jsx (hunk 10 above)
Comment (chrisforrette) [reply lib/commands/search.jsx]: Yea, I'm gonna change that `h()` to some JSX.

File: lib/components/search.jsx (hunk 11 above)
Comment (chrisforrette) [reply lib/components/search.jsx]: In [a separate thread](https://github.com/npm/tink/pull/11#discussion_r231722498) I was talking about changing the yargs options to pass down all the options of [`libnpmsearch`](https://www.npmjs.com/package/libnpmsearch), so this will just change to the default for the `limit` option, but allow it to be overridden by the user.

File: lib/commands/search.jsx (hunk 13 above)
Comment (chrisforrette) [reply lib/commands/search.jsx]: For now, I was experimenting so file exports and organization isn't _too_ clean just yet. But normally in frontendy React projects, I organize things like: `src/components/(component name)/index.js`, and `index.js` exports the component as the default, and I would also have `index.test.js`, or `index.stories.js` if [Storybook](https://www.npmjs.com/package/storybook) is being used, etc. At this point, that organization structure seems overkill as there aren't many React components just yet, but changing `module.exports = Search` will make it easy to switch to organizing things that way if components start to build up.

TL;DR: I'd like to make the export change 🙂

**Comment by chrisforrette** [comment]
@zkat Done!

**Comment by chrisforrette** [comment]
@zkat BTW, hoping to get to these changes this evening, sorry I've been a little slow...

**Comment by zkat** [comment]
You're fine! Take your time. We're not gonna be launching this until like next year so just take care of yourself first and foremost, and then if you have some time left over, go ahead and hack for fun <3

File: lib/components/search.jsx (hunk 6 above)
Comment (chrisforrette) [reply lib/components/search.jsx:217]: @zkat OK, so this update is made. The percentages feel a bit squished—I kinda wish there was a way to tab-align them so they aligned into a column—but it's definitely more clear. Gonna dive into `PackageView` integration soon...

File: lib/components/search.jsx (hunk 4 above)
Comment (chrisforrette) [reply lib/components/search.jsx]: @zkat I just updated this and tore `figures` and `ink-spinner` out since they have funky characters.

**Comment by chrisforrette** [comment]
@zkat Got a few updates besides the ones I've already commented on...

I've added an options object to the search command which is basically all of the `libnpmsearch` options, which are now being passed to the search call.

I am also trying to mimic your comment style now 😄 (I saw the CONTRIBUTING guide too late). Let me know if you'd like me go back and modify my commit messages.

Next I'm diving into that input focus thing—I have a good idea of user flow and might need to get a little hacky with it—and integrating the `PackageView`.

**Comment by zkat** [comment]
idea: Can we render things like a table? I think aligning the various elements will make it a lot more readable.

**Comment by larsgw** [comment]
I can make a `PackageSummary` to avoid obscuring the search select input and it jumping around given the relatively high and variable height of `PackageView`.

Maybe another command would be cool too, a full-screen (like `htop`/`tig`/`man`) terminal remake of the site. But that's definitely out of scope for this PR.

**Comment by chrisforrette** [comment]
@zkat Yea, I think rendering results in a table is a great idea. I'll see if I can compose components from [`ink-table`](http://npm.im/ink-table) and [`npm-select-input`](http://npm.im/ink-select-input) together to pull it off, but it might involve a custom franken-component.

@larsgw That sounds good—I assume you're thinking like a one-liner? I was originally thinking of putting it at the bottom of the matches list, but people might miss it if there are a lot of results.

**Comment by zkat** [comment]
How's it going with this?

**Comment by chrisforrette** [comment]
@zkat Good! Sorry for the radio silence.

I've been ironing out different states of searching to get them feeling right, and also being distracted by holiday busy-ness and such. After that I'm going to take a stab at displaying matching packages in a table view, but I'm thinking if it's less-than-straightforward, it might be better to put that into a follow-up PR if that's alright with you.

Same goes for the one-liner package view (displaying package info on one line when you highlight it). I'm not sure if this is something @larsgw is actively working on, or that I could/should work on, but it might be simpler in a separate PR.

**Comment by larsgw** [comment]
I haven't started on it yet, sorry for that. I was thinking of 3 lines of content, with maybe 2-3 lines of visual distinction ([I like this one](https://github.com/sindresorhus/ink-box)), but that's up to you.

**Comment by chrisforrette** [comment]
@larsgw No worries, you're not blocking me or anything. Let me know if you get started on it, and I'll do the same if I get to it first. Thanks!

**Comment by larsgw** [comment]
Just to make sure we don't do the same work twice: I made PR #38

**Comment by chrisforrette** [comment]
@larsgw Awesome, thanks!

**Comment by chrisforrette** [comment]
@zkat OK, I think this is ready for another look!

I think I ironed out the input focus stuff—I did a few tests with and without an initial term search, then deleting and entering a new search and it's feeling pretty good.

And I looked into using `ink-table` and it isn't a simple drop-in to use it for the selector. [There's some pretty unique logic](https://github.com/maticzav/ink-table/blob/master/src/index.js#L86) to render out the table, so I think to pull it off we would indeed need to make a franken-component out of it and `ink-select-input`. I'm down to build it, but I'd prefer to do it in a follow-up PR if that's cool.

And once #38 is good to go, I can drop it in over here.

Let me know what you think, thanks!

**Comment by zkat** [comment]
This looks great so far!

**Comment by chrisforrette** [comment]
@zkat Awesome, thanks!

For getting this merged, do you want the table display and the one-liner package description in place first?

**Comment by zkat** [comment]
@chrisforrette oh geez sorry it took so long to notice this again! No, I don't think table display is necessary yet, but I'd like to see more of a summary for each displayed package.

**Comment by chrisforrette** [comment]
@zkat No worries! I'll implement the `PackageSearchResult` component from @larsgw for the package summary once npm/tink#38 is merged.

**Comment by chrisforrette** [comment]
@zkat Btw, is there a real package installer I should use yet? I noticed [`lib/installer.js`](https://github.com/npm/tink/blob/latest/lib/installer.js) but it looks like that's installing everything from `package.json`. Should I add an `addPackage` method in there or something?

**Comment by SchneiderOr** [comment]
Hey guys any news about this feature? Would love to assist if any further help is needed

**Comment by chrisforrette** [comment]
@SchneiderOr I was waiting on #38 to implement here, I'll go nudge over there...

**Comment by chrisforrette** [comment]
@zkat Also I'm not sure if you saw my last comment, but is there a real install function I should wire up to this? Or should it remain nerfed?
